File: src/ValidateBinarySearchTree.java
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public boolean isValidBST(TreeNode root) {
        if (root == null) return true;
        Stack<TreeNode> st = new Stack<TreeNode>();
        int val = Integer.MIN_VALUE;
        while (!st.isEmpty() || root != null) {
            if (root != null) {
                st.push(root);
                root = root.left;
            } else {
                root = st.pop();
                if (root.val <= val) return false;
                val = root.val;
                root = root.right;
            }
        }
        return true;
    }
}

File: ValidateBinarySearchTree.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int flag;
    bool isValidBST(TreeNode *root) {
        flag = -100000;
        return isVBST(root);
    }
    
    bool isVBST(TreeNode *root) {
        if (!root) return true;
        if (!isVBST(root->left)) return false;
        if (flag >= root->val) return false;
        flag = root->val;
        return isVBST(root->right);
    }
};
